### PR TITLE
[rush] Remove the lodash dependency from rush-lib.

### PR DESCRIPTION
**Pull request by iclanton**
## Summary

Removes the dependency on `loadsh` from `@microsoft/rush-lib`. This should mitigate https://github.com/microsoft/rushstack/issues/4155

## How it was tested

Ran `rush update`, `rush install`, `rush upgrade-interactive`, `rush change`, and `rush build`.

### CODE REVIEW

File: libraries/rush-lib/src/utilities/objectUtilities.ts [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+          }
+        } else {
+          const aObjectProperties: Set<string> = new Set(Object.getOwnPropertyNames(a));
+          const bObjectProperties: Set<string> = new Set(Object.getOwnPropertyNames(b));
+          if (aObjectProperties.size !== bObjectProperties.size) {
+            return false;
+          } else {
+            for (const property of aObjectProperties) {
+              if (bObjectProperties.has(property)) {
+                bObjectProperties.delete(property);
```
Comment (dmichon-msft) [inline libraries/rush-lib/src/utilities/objectUtilities.ts:40]:
```suggestion
              if (bObjectProperties.delete(property)) {
```

File: libraries/rush-lib/src/utilities/test/objectUtilities.test.ts [hunk 2]
```diff
@@ -0,0 +1,166 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See the @microsoft/rush package's LICENSE file for license information.
+
+describe('objectUtilities', () => {
+  describe('objectsAreDeepEqual', () => {
+    it('can compare primitives', async () => {
+      const { objectsAreDeepEqual } = await import('../objectUtilities');
```
Comment (dmichon-msft) [inline libraries/rush-lib/src/utilities/test/objectUtilities.test.ts:7]: The use of `await import()` in unit tests is really only relevant for code that needs to have mocked dependencies, which this isn't. Making all the tests synchronous will likely improve test performance.

File: libraries/rush-lib/src/utilities/test/objectUtilities.test.ts [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+      const a: Record<string, number> = { a: 1 };
+      expect(cloneDeep({ a, b: a })).toEqual({ a: { a: 1 }, b: { a: 1 } });
+    });
+
+    it('can clone nested objects', async () => {
+      const { cloneDeep } = await import('../objectUtilities');
+
+      expect(cloneDeep({ a: { b: 1 } })).toEqual({ a: { b: 1 } });
+    });
```
Comment (dmichon-msft) [inline libraries/rush-lib/src/utilities/test/objectUtilities.test.ts:122]: All three of these tests should have checks that the resulting object is `!==` to the original.

**Review by dmichon-msft — APPROVED** [review]

**Review by D4N14L — APPROVED** [review]

File: libraries/rush-lib/src/utilities/objectUtilities.ts [hunk 4]
```diff
@@ -0,0 +1,117 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See the @microsoft/rush package's LICENSE file for license information.
+
+/**
+ * Determines if two objects are deeply equal.
+ */
+export function objectsAreDeepEqual<TObject>(a: TObject, b: TObject): boolean {
```
[the remark this answers is not in the document]
Comment (iclanton) [reply libraries/rush-lib/src/utilities/objectUtilities.ts:7]: I'll do that in a separate PR.

File: libraries/rush-lib/src/logic/installManager/InstallHelpers.ts [hunk 5]
```diff
@@ -3,26 +3,26 @@
...
 import { IConfigurationEnvironment } from '../base/BasePackageManagerOptionsConfiguration';
-
-const lodash: typeof import('lodash') = Import.lazy('lodash', require);
+import type { PnpmOptionsConfiguration } from '../pnpm/PnpmOptionsConfiguration';
+import { merge } from '../../utilities/objectUtilities';
 
 interface ICommonPackageJson extends IPackageJson {
-  pnpm?: unknown;
+  pnpm?: {
+    overrides?: typeof PnpmOptionsConfiguration.prototype.globalOverrides;
```
[the remark this answers is not in the document]
Comment (iclanton) [reply libraries/rush-lib/src/logic/installManager/InstallHelpers.ts:19]: As more things get added, let's just add them explicitly here.

**Pull request merged**